### PR TITLE
Sahithi/stalwart-162

### PR DESCRIPTION
**Pull request by SahithiMy**
### :nut_and_bolt: Description: What code changed, and why?
  Added a function regarding the day_latest and daily_latest flag, Based on the range this returns the day latest or daily latest.

### :chains: Related Resources
https://chefio.atlassian.net/browse/STALWART-162

### :+1: Definition of Done
Changes have made for the flags day_latest and daily_latest

### :athletic_shoe: How to Build and Test the Change
curl --location --request POST 'https://ec2-3-22-57-109.us-east-2.compute.amazonaws.com/api/v0/compliance/reporting/nodes/search' \
--header 'api-token: 3R96rLWzFGlzOTnjh058zVHPEwA=' \
--header 'Content-Type: application/json' \
--data-raw '{
    "filters": [
        {
            "type": "start_time",
            "values": [
                "2022-04-20T00:00:00Z"
            ]
        },
        {
            "type": "end_time",
            "values": [
                "2022-06-23T23:59:59Z"
            ]
        }
    ]
}'

### :white_check_mark: Checklist

**All PRs** must tick these:

- [X] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [X] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [X] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [X] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [X] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [X] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [X] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [X] Tests added/updated? (all new code needs new tests)
- [X] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: components/compliance-service/reporting/relaxting/reports.go [hunk 1]
```diff
@@ -1163,14 +1192,28 @@ func (backend ES2Backend) getFiltersQuery(filters map[string][]string, latestOnl
 
 	if latestOnly {
 		// only if there is no job_id filter set, do we want the daily latest
+		s, err := filterQuerychange(filters)
+		if err != nil {
+			panic(err)
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1197]: We cannot panic in code. The error needs to be handled

File: components/compliance-service/reporting/relaxting/reports.go [hunk 2]
```diff
@@ -1163,14 +1192,28 @@ func (backend ES2Backend) getFiltersQuery(filters map[string][]string, latestOnl
...
+		}
 		if len(filters["end_time"]) > 0 {
-			// If we have an end_time filter, we use the daily_latest filter dedicated to the UTC timeseries indices
-			termQuery := elastic.NewTermsQuery("daily_latest", true)
-			boolQuery = boolQuery.Must(termQuery)
+			if len(s) == 1 && s[0] == "daily_latest" {
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1200]: This complete code can be moduled out to a different function or at least a loop.

```
for search_field := range s {
      termQuery := elastic.NewTermsQuery(search_field, true)
      boolQuery = boolQuery.Must(term query)
}
```

File: components/compliance-service/reporting/relaxting/reports_test.go [hunk 3]
```diff
@@ -206,3 +206,20 @@ func TestDoesControlTagMatchFilter(t *testing.T) {
 	multFilters["control_tag:shoe"] = []string{"blue"}
 	assert.Equal(t, true, doesControlTagMatchFilter(multFilters, tags2))
 }
+
+func TestFilterQuerychange(t *testing.T) {
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports_test.go:210]: Success, failure and corner cases need to tested. This is not a good way of writing unit tests.

File: components/compliance-service/reporting/relaxting/reports.go [hunk 4]
```diff
@@ -1123,7 +1139,20 @@ func (backend ES2Backend) getFiltersQuery(filters map[string][]string, latestOnl
...
+		layout := "2006-01-02T15:04:05Z"
+		endtime, err := time.Parse(layout, endTime)
+		starttime, err := time.Parse(layout, startTime)
+		if err != nil {
+			panic(err)
+		}
+		diff := endtime.Sub(starttime).Hours() / 24
 
+		if diff > 90 {
+			logrus.Error("Range should be less than 90 days")
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1151]: The error should not only be logged but also responded back to the API. We cannot continue if the diff is more than 90 days

File: components/compliance-service/reporting/relaxting/reports.go [hunk 5]
```diff
@@ -1123,7 +1139,20 @@ func (backend ES2Backend) getFiltersQuery(filters map[string][]string, latestOnl
...
+		}
+		if diff < 0 {
+			logrus.Error("Start date should not be greater than end date")
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1154]: Same her

File: components/compliance-service/reporting/relaxting/reports.go [hunk 6]
```diff
@@ -1123,7 +1139,20 @@ func (backend ES2Backend) getFiltersQuery(filters map[string][]string, latestOnl
 	if len(filters["start_time"]) > 0 || len(filters["end_time"]) > 0 {
 		endTime := firstOrEmpty(filters["end_time"])
 		startTime := firstOrEmpty(filters["start_time"])
+		layout := "2006-01-02T15:04:05Z"
+		endtime, err := time.Parse(layout, endTime)
+		starttime, err := time.Parse(layout, startTime)
+		if err != nil {
+			panic(err)
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1146]: Cannot panic!!!!!

File: components/compliance-service/reporting/relaxting/reports.go [hunk 7]
```diff
@@ -1123,7 +1139,20 @@ func (backend ES2Backend) getFiltersQuery(filters map[string][]string, latestOnl
...
+		}
+		diff := endtime.Sub(starttime).Hours() / 24
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1148]:
```suggestion
		diff := int(endtime.Sub(starttime).Hours() / 24)
```

File: components/compliance-service/reporting/relaxting/reports.go [hunk 8]
```diff
@@ -1063,6 +1063,22 @@ func (backend *ES2Backend) getWaiverData(waiverDataBuckets *elastic.AggregationB
...
+	layout := "2006-01-02T15:04:05Z"
+	endtime, err := time.Parse(layout, endTime)
+	starttime, err := time.Parse(layout, startTime)
+	if err != nil {
+		panic(err)
+	}
+	diff := endtime.Sub(starttime).Hours() / 24
+	if diff == 0 {
+		return []string{"daily_latest"}, nil
+	} else {
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1078]: Unnecessary `else` block

File: components/compliance-service/reporting/relaxting/reports.go [hunk 9]
```diff
@@ -1063,6 +1063,22 @@ func (backend *ES2Backend) getWaiverData(waiverDataBuckets *elastic.AggregationB
 
 	return waiverDataCollection, nil
 }
+func filterQuerychange(filters map[string][]string) ([]string, error) {
+	endTime := firstOrEmpty(filters["end_time"])
+	startTime := firstOrEmpty(filters["start_time"])
+	layout := "2006-01-02T15:04:05Z"
+	endtime, err := time.Parse(layout, endTime)
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1070]: Follow a single naming pattern and it has to be `camelCase`

File: components/compliance-service/reporting/relaxting/reports.go [hunk 10]
```diff
@@ -1063,6 +1063,22 @@ func (backend *ES2Backend) getWaiverData(waiverDataBuckets *elastic.AggregationB
 
 	return waiverDataCollection, nil
 }
+func filterQuerychange(filters map[string][]string) ([]string, error) {
+	endTime := firstOrEmpty(filters["end_time"])
+	startTime := firstOrEmpty(filters["start_time"])
+	layout := "2006-01-02T15:04:05Z"
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1069]: No need of the layout. Should be put directly on the Parse function

**Review by kalroy — CHANGES_REQUESTED** [review]

**Review by kalroy — APPROVED** [review]

File: components/compliance-service/reporting/relaxting/reports_test.go [hunk 11]
```diff
@@ -206,3 +206,20 @@ func TestDoesControlTagMatchFilter(t *testing.T) {
...
+	filters := map[string][]string{}
+	filters["start_time"] = []string{"2022-06-22T00:00:00Z"}
+	filters["end_time"] = []string{"2022-06-23T00:00:00Z"}
+	s, err := filterQuerychange(filters)
```
Comment (dave-augustus) [inline components/compliance-service/reporting/relaxting/reports_test.go:214]: give s a meaningful name.

File: components/compliance-service/reporting/relaxting/reports_test.go [hunk 12]
```diff
@@ -206,3 +206,20 @@ func TestDoesControlTagMatchFilter(t *testing.T) {
...
+	if err != nil {
+		return
+	}
+	assert.Equal(t, []string{"day_latest", "daily_latest"}, s)
+	filters1 := map[string][]string{}
+	filters1["start_time"] = []string{"2022-06-21T00:00:00Z"}
+	filters1["end_time"] = []string{"2022-06-21T00:00:00Z"}
+	a, err := filterQuerychange(filters1)
```
Comment (dave-augustus) [inline components/compliance-service/reporting/relaxting/reports_test.go:222]: give 'a' a meaningful name

**Review by YashviJain01 — APPROVED** [review]

**Review by dave-augustus — APPROVED** [review]

File: components/compliance-service/reporting/relaxting/reports.go [hunk 13]
```diff
@@ -1325,7 +1322,15 @@ func (backend ES2Backend) getFiltersQuery(filters map[string][]string, latestOnl
...
 		startTime := firstOrEmpty(filters["start_time"])
-
+		diff, _, _ := filterQuerychange(endTime, startTime)
+		if diff > 90 {
+			logrus.Error("Range should be less than 90 days")
+			return nil
+		}
+		if diff < 0 {
+			logrus.Error("Start date should not be greater than end date")
+			return nil
```
Comment (kalroy) [inline components/compliance-service/reporting/relaxting/reports.go:1332]: If we return nil, how will the API respond back to the user with the message?

**Review by kalroy — CHANGES_REQUESTED** [review]

**Review by dave-augustus — APPROVED** [review]
